consensus/src/test/kotlin/maru/consensus/validation/BlockValidatorImplTest.kt

```diff
@@ -0,0 +1,257 @@
+
+import maru.consensus.ValidatorProvider
+import maru.core.BeaconBlock
+import maru.core.ExecutionPayload
+import maru.core.HashUtil
+import maru.core.Seal
+import maru.core.SealedBeaconBlock
+import maru.core.Validator
+import maru.core.ext.DataGenerators
+import maru.database.Database
+import maru.executionlayer.client.ExecutionLayerClient
+import maru.serialization.rlp.bodyRoot
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import org.mockito.kotlin.any
+import org.mockito.kotlin.argThat
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.eq
+import org.mockito.kotlin.mock
+import tech.pegasys.teku.ethereum.executionclient.schema.PayloadStatusV1
+import tech.pegasys.teku.ethereum.executionclient.schema.Response
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+import tech.pegasys.teku.spec.executionlayer.ExecutionPayloadStatus
+
+class BlockValidatorImplTest {
+  private val validPrevBlockBody = DataGenerators.randomBeaconBlockBody()
+  private val validPrevBlockHeader =
+    DataGenerators.randomBeaconBlockHeader(9u).copy(
+      bodyRoot = HashUtil.bodyRoot(validPrevBlockBody),
+    )
+  private val validPrevBlock = BeaconBlock(validPrevBlockHeader, validPrevBlockBody)
+  private val validBlockBody = DataGenerators.randomBeaconBlockBody()
+  private val validBlockHeader =
+    DataGenerators.randomBeaconBlockHeader(validPrevBlockHeader.number + 1u).copy(
+      parentRoot = validPrevBlockHeader.hash,
+      timestamp = validPrevBlockHeader.timestamp + 1u,
+      bodyRoot = HashUtil.bodyRoot(validBlockBody),
+    )
+  private val validBlock = BeaconBlock(validBlockHeader, validBlockBody)
+
+  private val validatorProvider =
+    mock<ValidatorProvider> {
+      on { getProposerForBlock(any()) }.thenReturn(validBlockHeader.proposer)
+      on { getValidatorsForBlock(any()) }
+        .thenReturn((0 until validBlockBody.prevCommitSeals.size).map { DataGenerators.randomValidator() }.toSet())
+    }
+  private val sealVerifier =
+    mock<SealVerifier> {
+      on { verifySeal(any(), any()) } doReturn (Result.failure(RuntimeException("Invalid seal")))
+      on {
+        verifySeal(
+          argThat { seal: Seal ->
+            validBlockBody.prevCommitSeals.contains(seal)
+          },
+          eq(validPrevBlockHeader),
+        )
+      } doReturn (Result.success(true))
+    }
+
+  private val executionLayerClient =
+    mock<ExecutionLayerClient> {
+      on { newPayload(any()) }.thenAnswer {
+        val executionPayload = it.getArgument<ExecutionPayload>(0)
+        if (executionPayload == validBlockBody.executionPayload) {
+          val mockPayload = PayloadStatusV1(ExecutionPayloadStatus.VALID, null, null)
+          SafeFuture.completedFuture(Response(mockPayload))
+        } else {
+          SafeFuture.completedFuture(Response.withErrorMessage<PayloadStatusV1>("Invalid execution payload"))
+        }
+      }
+    }
+
+  private val database =
```

Comment:
I have the same suggestion here as for the `executionLayerClient`. It would be better to make the mock unconditional by default and make it overridable for the negative tests.